Moq.AutoMock/AutoMocker.cs

```diff
@@ -32,14 +33,45 @@ public AutoMocker(MockBehavior mockBehavior, DefaultValue defaultValue)
 
         public AutoMocker(MockBehavior mockBehavior, DefaultValue defaultValue, bool callBase)
         {
-            this.MockBehavior = mockBehavior;
-            this.DefaultValue = defaultValue;
-            this.CallBase = callBase;
+            MockBehavior = mockBehavior;
+            Resolvers.Add(new MoqResolver(mockBehavior, defaultValue, callBase));
         }
 
-        public MockBehavior MockBehavior { get; private set; }
-        public DefaultValue DefaultValue { get; private set; }
-        public bool CallBase { get; private set; }
+        public MockBehavior MockBehavior { get; }
+        public ICollection<IMockResolver> Resolvers { get; } = new List<IMockResolver>();
+
+        private IInstance Resolve(Type serviceType)
+        {
+            if (serviceType.IsArray)
+            {
+                Type elmType = serviceType.GetElementType();
+                MockArrayInstance instance = new MockArrayInstance(elmType);
+                if (typeMap.TryGetValue(elmType, out var element))
+                    instance.Add(element);
+                return instance;
+            }
+
+            var context = new MockResolutionContext
+            {
+                AutoMocker = this,
+                RequestType = serviceType,
+                Value = null //or default to a Mock<T>? MoqResolver is our first Resolver and will create one
+            };
+
+            foreach (var r in Resolvers)
+                r.Resolve(context);

```

Comment:
This pattern of creating a context and iterating over all of the resolves is repeated several times. Move it into a single method.